pkg/webhooks/validation/controller/controller.go

```diff
@@ -0,0 +1,530 @@
+		ownerRefs:     findClusterRoleOwnerRefs(o.Client, o.ClusterRoleName),
+	}
+
+	c.sharedInformers = informers.NewSharedInformerFactoryWithOptions(o.Client, o.ResyncPeriod,
+		informers.WithNamespace(o.WatchedNamespace))
+
+	webhookInformer := c.sharedInformers.Admissionregistration().V1beta1().ValidatingWebhookConfigurations().Informer()
+	webhookInformer.AddEventHandler(makeHandler(c.queue, configGVK, o.WebhookConfigName))
+
+	endpointInformer := c.sharedInformers.Core().V1().Endpoints().Informer()
+	endpointInformer.AddEventHandler(makeHandler(c.queue, endpointGVK, o.ServiceName))
+
+	deploymentInformer := c.sharedInformers.Apps().V1().Deployments().Informer()
+	deploymentInformer.AddEventHandler(makeHandler(c.queue, deploymentGVK, o.GalleyDeploymentName))
+
+	return c, nil
+}
+
+func (c *Controller) Start(stop <-chan struct{}) {
+	go c.startFileWatcher(stop)
+	go c.sharedInformers.Start(stop)
+
+	for _, ready := range c.sharedInformers.WaitForCacheSync(stop) {
+		if !ready {
+			return
+		}
+	}
+
+	req := &reconcileRequest{"initial request to kickstart reconciliation"}
+	c.queue.Add(req)
+
+	go c.runWorker()
+}
+
+func (c *Controller) startFileWatcher(stop <-chan struct{}) {
+	for {
+		select {
+		case ev := <-c.caFileWatcher.Events(c.o.WebhookConfigPath):
+			req := &reconcileRequest{fmt.Sprintf("Webhook configuration changed: %v", ev)}
+			c.queue.Add(req)
+		case ev := <-c.caFileWatcher.Events(c.o.CAPath):
+			req := &reconcileRequest{fmt.Sprintf("CA file changed: %v", ev)}
+			c.queue.Add(req)
+		case err := <-c.caFileWatcher.Errors(c.o.WebhookConfigPath):
+			scope.Warnf("error watching local webhook configuration: %v", err)
+		case err := <-c.caFileWatcher.Errors(c.o.CAPath):
+			scope.Warnf("error watching local CA bundle: %v", err)
+		case <-stop:
+			return
+		}
+	}
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() (cont bool) {
+	obj, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	req, ok := obj.(*reconcileRequest)
+	if !ok {
+		// don't retry an invalid reconcileRequest item
+		c.queue.Forget(req)
+		return true
+	}
+
+	if err := c.reconcileRequest(req); err != nil {
+		c.queue.AddRateLimited(obj)
+		utilruntime.HandleError(err)
+	} else {
+		c.queue.Forget(obj)
+	}
+	return true
+}
+
+// reconcile the desired state with the kube-apiserver.
+func (c *Controller) reconcileRequest(req *reconcileRequest) error {
+	defer func() {
+		if c.reconcileDone != nil {
+			c.reconcileDone()
+		}
+	}()
+
+	scope.Infof("Reconcile: %v", req)
+
+	// don't create the webhook config before the endpoint is ready
+	if !c.endpointReadyOnce {
+		if ready, err := c.isEndpointReady(); !ready || err != nil {
+			scope.Infof("Endpoint not ready: ready=%v err=%v", ready, err)
+			return err
+		}
+		c.endpointReadyOnce = true
+	}
+
+	// don't update the webhook config if its already managed by an existing galley deployment.
```

Comment:
With the sidecar injector we just had some logic in the install that

`if istiod then pilot reconciles, else injector reconciles`

Are there benefits to this approach?

Reply:
Injection and validation have slightly different requirements for install/upgrade. 

The injection config can be created before the injection webhook is ready. Pod creation will fail but the higher-level controller will retry. This is also less of a problem because we don't typically install user apps concurrently with installing Istio. 

The validation config can only be created after the validation webhook is ready. Otherwise, the  Istio CRs included in the default install may be rejected. The sequence of waiting for endpoint readiness before config creation avoids this problem. This is why the validatingwebhookconfiguration is created by Galley and not included in the istio.yaml.

> if istiod then pilot reconciles, else injector reconciles
>
> Are there benefits to this approach?

The current injection logic is good. We shouldn't add the more complicated validation logic if its not needed.